clang/lib/DPCT/RulesSHMEM/NVSHMEMAPIMigration.cpp

```diff
@@ -123,64 +249,99 @@ void clang::dpct::NVSHMEMRule::runRule(
         }
 
         // Get the second argument as string
-        attr_arg = Lexer::getSourceText(
-            CharSourceRange::getTokenRange(CE->getArg(1)->getSourceRange()),
-            DpctGlobalInfo::getSourceManager(), LangOptions());
+        AttrArg = Lexer::getSourceText(
+            CharSourceRange::getTokenRange(CE->getArg(1)->getSourceRange()), SM,
+            LO);
 
-        if (ishmem_rt.empty()) {
+        if (IshmemRT.empty()) {
           report(CE->getBeginLoc(), Diagnostics::API_NOT_MIGRATED, false,
                  FuncName);
           return;
         } else {
-          auto &SM = DpctGlobalInfo::getSourceManager();
-
           auto IndentLoc = CE->getBeginLoc();
           if (IndentLoc.isMacroID())
             IndentLoc = SM.getExpansionLoc(IndentLoc);
 
-          std::string set_ishmem_runtime =
-              "(" + attr_arg + ")->runtime = " + ishmem_rt + ";";
-          set_ishmem_runtime += getNL();
-          set_ishmem_runtime += getIndent(IndentLoc, SM).str();
+          std::string SetIshmemRT =
+              "(" + AttrArg + ")->runtime = " + IshmemRT + ";";
+          SetIshmemRT += getNL();
+          SetIshmemRT += getIndent(IndentLoc, SM).str();
 
           emplaceTransformation(
-              new InsertBeforeStmt(CE, std::move(set_ishmem_runtime)));
+              new InsertBeforeStmt(CE, std::move(SetIshmemRT)));
+        }
+      } else if (FuncName == "nvshmemx_signal_op") {
+        std::string IshmemxSignalOP = "ishmemx_signal_";
+        std::string SigOPInit = "";
+
+        // Get third argument's init value
+        auto Arg = CE->getArg(2);
+        if (auto DRE = dyn_cast<DeclRefExpr>(Arg->IgnoreImpCasts())) {
+          if (const VarDecl *VD = dyn_cast<VarDecl>(DRE->getDecl())) {
+            if (VD->hasInit()) {
+              // get the init value from definition
+              if (auto Init =
+                      dyn_cast<DeclRefExpr>(VD->getInit()->IgnoreImplicit())) {
+                SigOPInit = Init->getNameInfo().getName().getAsString();
+              }
+            }
+          }
+        }
+
+        if (SigOPInit == "NVSHMEM_SIGNAL_SET") {
+          IshmemxSignalOP += "set";
+        } else if (SigOPInit == "NVSHMEM_SIGNAL_ADD") {
+          IshmemxSignalOP += "add";
+        } else {
+          report(CE->getBeginLoc(), Diagnostics::API_NOT_MIGRATED, false,
+                 FuncName);
         }
+
+        // Removed third arg
+        auto LocAfterArg = Lexer::findLocationAfterToken(
+            Arg->getEndLoc(), tok::comma, SM, LO, true);
+        emplaceTransformation(
+            new ReplaceText(Arg->getBeginLoc(), LocAfterArg, std::move("")));
+
+        // Replace function name
+        emplaceTransformation(new ReplaceText(CE->getCallee()->getBeginLoc(),
+                                              FuncName.length(),
+                                              std::move(IshmemxSignalOP)));

```

Comment:
I suggest to use `getDefinitionRange` to get the range of this `CallExpr` and replace the whole `CallExpr` together.

Reply:
Included the new helper function to handle this API migration